New ShortCode for iDonate Fund Selector

blairlierman opened the pull request:
PR for Issue #4: Created a shortcode in WordPress named "idonate_fundselector" that displays the necessary HTML to allow a donor to select a fund from a list of categories or search through all funds.

The current state of the short code will show the funds that are categorized as priorities and also allows the user to search through all funds using the jQuery UI Autocomplete control combined with an WordPress AJAX method.

Currently there is no styling on the controls other than default. The addition of more styling is referenced in issue #10.

jeremyfelt commented on includes/class-ajax-data-retrieval.php:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	public function wsuwp_plugin_idonate_ajax_fund_search() {
+		$search_term = strtolower( $_REQUEST['term'] );
```
When using `wp_ajax_`, there should be a nonce to make sure the request is coming from the right place, especially when `wp_ajax_nopriv` is in use as that can allow any unauthenticated user to make a request. See [this](https://github.com/washingtonstateuniversity/project.wsu.edu/blob/657424055f96cccbf144f16bfdd6895b8d360ddb/functions.php#L91), [this](https://github.com/washingtonstateuniversity/project.wsu.edu/blob/657424055f96cccbf144f16bfdd6895b8d360ddb/functions.php#L66), and [this](https://github.com/washingtonstateuniversity/project.wsu.edu/blob/657424055f96cccbf144f16bfdd6895b8d360ddb/js/project-create.js#L24) as a loose example of the required pieces.

jeremyfelt commented on includes/class-ajax-data-retrieval.php:
```diff
@@ -0,0 +1,62 @@
+	public function init() {
+
+		add_action( 'wp_ajax_nopriv_wsuwp_plugin_idonate_ajax_fund_search', array( $this, 'wsuwp_plugin_idonate_ajax_fund_search' ) );
+		add_action( 'wp_ajax_wsuwp_plugin_idonate_ajax_fund_search', array( $this, 'wsuwp_plugin_idonate_ajax_fund_search' ) );
```
`wp_ajax` will work for this, but you may want to consider the ease of the WordPress REST API, which is available on all sites on the platform. Post types can be [exposed](https://github.com/washingtonstateuniversity/WSU-25x2030-Theme/blob/master/includes/class-drive-story-post-type.php#L77) in the API and then [consumed in JS](https://github.com/washingtonstateuniversity/WSU-25x2030-Theme/blob/master/js/evidence.js#L93) via that post type's [endpoint](https://github.com/washingtonstateuniversity/WSU-25x2030-Theme/blob/2d4d085cdf81124192b7117516e11e8a2d395fa4/functions.php#L108).

That's not necessarily the way to go, as it requires the post type is public, but may be the easiest as far as handling the results in JS. That said, with the admin ajax approach already implemented, definitely not a requirement.

blairlierman replied:
@jeremyfelt We are considering moving to the REST API (which I didn't realize was an option). Are there any other caveats to using it aside from making the post type public?

jeremyfelt commented on includes/class-ajax-data-retrieval.php:
```diff
@@ -0,0 +1,62 @@
+				$title = get_the_title();
+				$des_id = get_post_meta( get_the_ID(), 'designationId', true );
+
+				$return[] = array( 'label' => $title , 'name' => $title, 'designationId' => $des_id );
```
Make sure that `$des_id` is escaped properly before being returned. `wp_json_encode` later on may help a bit, but it'd be good to sanitize it so that it's a value that's expected. I'm not sure if that would be `absint()` or `sanitize_key()` or something similar.

jeremyfelt commented on includes/class-wsuwp-shortcode-fundselector.php:
```diff
@@ -0,0 +1,167 @@
+		foreach ( $priorities as $priority ) {
+			$fund_name = $priority['fund_name'];
+			$fund_designation_id = $priority['designation_id'];
+			$priorities_list .= "<option value=\"{$fund_designation_id}\">{$fund_name}</option>";
```
Use `esc_attr()` when outputting data like `$fund_designation_id` into an HTML attribute. `esc_html()` should work (unless you're allowing HTML tags) for `$fund_name`. In both cases, the data should be escaped before output.

blairlierman commented:
@jeremyfelt Thank you for the review. The REST API appeared to be the way to go for this, so I switched to using it. I also added in the HTML data escaping.

blairlierman commented:
@jeremyfelt Any thoughts on the updated changes to this PR?
